FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageQueueClient.cs

```diff
@@ -0,0 +1,682 @@
+                {
+                    _logger.LogWarning("Failed to keep alive, the job message is not found.");
+                }
+            }
+
+            // step 5: sync result to jobInfo entity
+            // if update message successfully while update table entity failed, then the message pop receipt is invalid,
+            // keeping alive always fails to update message, so the message will be visible and dequeue again
+            // when re-dequeue, the message pop receipt is updated to table entity,
+            // and the previous job will throw JobNotExistException as the version doesn't match, and jobHosting cancels the previous job.
+            jobInfoEntity.HeartbeatDateTime = DateTime.UtcNow;
+
+            Trace.Assert(
+                jobInfo.Result.Length * sizeof(char) < MaximumSizeOfAnEntityInBytes,
+                "The maximum size of a single table entity is 1MB, the size of result is larger than 1MB.");
+
+            jobInfoEntity.Result = jobInfo.Result;
+
+            // step 6: update message pop receipt to job lock entity
+            jobLockEntity.JobMessagePopReceipt = response?.Value.PopReceipt;
+
+            // step 7: transaction update jobInfo entity and job lock entity
+            IEnumerable<TableTransactionAction> transactionUpdateActions = new List<TableTransactionAction>
+            {
+                new (TableTransactionActionType.UpdateReplace, jobInfoEntity, jobInfoEntity.ETag),
+                new (TableTransactionActionType.UpdateReplace, jobLockEntity, jobLockEntity.ETag),
+            };
+
+            _ = await _azureJobInfoTableClient.SubmitTransactionAsync(transactionUpdateActions, cancellationToken);
+
+            // step 8: check if cancel requested
+            var shouldCancel = (await GetJobByIdAsync(jobInfo.QueueType, jobInfo.Id, false, cancellationToken)).CancelRequested;
+
+            _logger.LogInformation($"Keep alive for job {jobInfo.Id} successfully.");
+
+            return shouldCancel;
+        }
+
+        public async Task CancelJobByGroupIdAsync(byte queueType, long groupId, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation($"Start to cancel jobs in group {groupId}.");
+
+            var jobInfoEntities = new List<JobInfoEntity>();
+
+            // step 1: query all job ids in a group, using range query for row key to ignore lock/index entities in same partition
+            var queryResult = _azureJobInfoTableClient.QueryAsync<JobInfoEntity>(
+                filter: FilterJobInfosByGroupId(queueType, groupId),
+                cancellationToken: cancellationToken);
+            await foreach (var pageResult in queryResult.AsPages().WithCancellation(cancellationToken))
+            {
+                foreach (var jobInfoEntity in pageResult.Values)
+                {
+                    // step 2: cancel job.
+                    CancelJobInternal(jobInfoEntity);
+                    jobInfoEntities.Add(jobInfoEntity);
```

Comment:
Do we need to update the `jobinfo` that status is not cancelled here? Maybe we can only update status cancelled jobs?

Reply:
This function is called when a job failed and requires to cancel other jobs in the same group.
So we need to cancel all the jobs in this group, we just set the `CancelRequst` in the entity to true and don't set the status to canceled at this stage, which is as what `CancelJobByIdAsync` do.